core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +75,225 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
                 .build();
         }
 
-        SdkHttpRequest signedRequest = doSign(request).build();
+        ContentChecksum contentChecksum = createChecksum(request);
+
+        SdkHttpRequest.Builder requestBuilder = sign(request, contentChecksum);
+
         return SyncSignedRequest.builder()
-            .request(signedRequest)
+            .request(requestBuilder.build())
             .payload(request.payload().orElse(null))
             .build();
     }
 
-    private SdkHttpRequest.Builder doSign(SyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        String checksumHeaderName = request.property(CHECKSUM_HEADER_NAME);
-        ChecksumAlgorithm checksumAlgorithm = request.property(CHECKSUM_ALGORITHM);
-
-        if (StringUtils.isNotBlank(checksumHeaderName) && checksumAlgorithm == null) {
-            throw new IllegalArgumentException(
-                CHECKSUM_ALGORITHM + " cannot be null when " + CHECKSUM_HEADER_NAME + " is given!");
+    @Override
+    public AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request) {
```

Comment:
In making this class non-final, again to start with strictness and relax later iff needed, make the public sign/signAsync methods final and the methods that don't need to be overridden (except the 3 in query signer) private.